### DISCUSSION TITLE
W1209 module with 5V

### DISCUSSION THREAD

**Discussion by scientistnobee** [TG9541/stm8ef#399] [Q&A]
Hi, I would like to make a small heat pad with 5V battery powerbank using W1209. However, all I found is 12 V modules. The 5V modules advertised in ebay/Aliexpress don't seem to have genuine parts. Could you please advise me how to get the genuine 5V module or how to convert a 12V module into a 5V module. I have seen a similar question in electronics stack exchange, but there is no satisfactory answer is given, particularly about MOSFET and relay.

https://electronics.stackexchange.com/questions/544292/can-w1209-be-run-with-5v-dc-only

**Comment by TG9541** [comment]
I had a look at the infos on stackexchange: 

* when the USB powerpack supply voltage is in the 4.5 to 5.5V range connecting the anode of D2 to the output of the 5V regulator should work. 

* After removing the 12V relay soldering a through-hole STP36NF06L at the same spot should work. It's maybe necessary to remove R7 so that the STM8S003 pin10 (PA3) doesn't get any additional load.

**Reply by krikoman** [reply]
Hi, I'm new to all this and know very little about electronics and flashing :) etc. 
I'm a electrical engineer with some C++ skills and that's about it.
I recently bought 5V W1209, but not sure if it's the sort you can flash / program whatever it's called)
I got it here https://www.ebay.co.uk/vod/FetchOrderDetails?itemid=293138632724&transid=1881944225019&ul_noapp=true
It has a chip labelled as CMS8S588 2015CN, I can't find any details on this anywhere.
I've been looking at this site, because I like the idea of logging temps and relay actions. But it all looks a bit scary and a bit beyond me.
Anyhow, thanks for all the info, I'll might give it a go some time.
Not sure if what I've posted is any use. let me know if you need more details, or if you can help on the chip ID of mine.
I bought another one not long after  a 12V with a case, really only needed the case. The chip on that one has no markings and is quite different from the picture on these pages.

Cheers,
Mark

**Reply by krikoman** [reply]
Oh! meant to say, the terminals are still marked as +12V and GND, but it works on 5V and the relay is 5V, I'm using it to heat a butter dish up using a USB cup warmer (who knew). Our kitchen gets cold and I'm sick of tearing bread to pieces trying to butter it.

**Reply by scientistnobee** [reply]
I couldn't find a link to the item on ebay. Mark, do you mind sharing the link and how long does it took to arrive. I have been looking for a 5V module since long.

**Reply by krikoman** [reply]
Try this, I thought the link above might have worked.
https://www.ebay.co.uk/itm/293138632724
It took 10 days, it was supposed to take 26 days longer, so I was pleasantly surprised.

**Reply by krikoman** [reply]
Just to add, the 12V version will work down to 4.5V, but the relay won't pull in, it should be easy enough to swap out the relay if you're desperate though.

**Reply by scientistnobee** [reply]
I am trying to replace the replay. However, my heating element has only 3.6 Ohms, so more than an 1 amp is flowing through the relay. That is upsetting the relay and it is clicking continuously, and not holding the connection.

**Reply by krikoman** [reply]
Most relays are good for at least 5A but either way the relay would simple burn out rather than drop in and out.
Are you powering your heater from the same power supply as the W1209? If this is the case, then probably what's happening is the relay switches, the load is switched and the voltage drops, the relay drop out, and drops the load, and it then cycles.
What size power supply are you using, do you have drawing of the circuit.
Also, depending on you heater element, there can be a massive inrush current, until it gets warmed up, a large load will pull the voltage down.

**Reply by TG9541** [reply]
Interesting discussions are going on here - there seems to be a real need for hackable thermostats ;-)
@krikoman I've never heard of the CMS8S588 but it sounds like yet another STM8S003F3-look-alike. Can you post a photo with a close-up of the MCU? Thanks!

How much would you (or @scientistnobee) be willing to pay for a hackable thermostat and which features would be important to you?

**Reply by scientistnobee** [reply]
> Most relays are good for at least 5A but either way the relay would simple burn out rather than drop in and out.
> Are you powering your heater from the same power supply as the W1209? If this is the case, then probably what's happening is the relay switches, the load is switched and the voltage drops, the relay drop out, and drops the load, and it then cycles.
> What size power supply are you using, do you have drawing of the circuit.
> Also, depending on you heater element, there can be a massive inrush current, until it gets warmed up, a large load will pull the voltage down.

Yes, I am powering W1209 with the same source. My heater is having 3.6 Ohms. My power supply is a 5V battery bank. So the current draw is more than one ampere. The power bank's voltage is dropping to 3.6V when relay is on. So it's struggling. Is there anyway to solve this issue? Like shorting the diode or the resistor before the voltage regulator.

**Reply by scientistnobee** [reply]
> Interesting discussions are going on here - there seems to be a real need for hackable thermostats ;-)
> @krikoman I've never heard of the CMS8S588 but it sounds like yet another STM8S003F3-look-alike. Can you post a photo with a close-up of the MCU? Thanks!
> 
> How much would you (or @scientistnobee) be willing to pay for a hackable thermostat and which features would be important to you?

Hi @TG9541 I am in Ireland. Please let me know if you have any genuine 5v modules. I wanna buy. I ordered one from China but I don't think it will come anytime soon. I am also thinking of designing a board for this in Eagle and get a few boards made through JLCPCB.

**Reply by TG9541** [reply]
@scientistnobee sure, I can send a W1209 from Germany to Ireland. I can remove the relay if you like, then I can send it in a letter. 
Regarding designing something and sourcing from JLCPCB (that works world-wide): 
* the power supply should be given a serious thought as @krikoman rightly pointed out
* applying a MOS-FET would make it possible to work with some kind of continuous mode (PWM) instead of relay and hysteresis 
* if you agree to using KiCad instead of Eagle then I can contribute to the design of a board (I've been thinking of doing that for a long time)

Here is something in my hobby pipeline that waits to be finished (to give you an idea): 
![7699501609782797307](https://user-images.githubusercontent.com/5466977/107870054-7fe79d00-6e95-11eb-9d43-8bc06a3fac65.png)

If you like the idea you can start with plotting down some specs here :-)

**Reply by scientistnobee** [reply]
Here is my wish list:

1. Should work with 5V
2. High precision (0.1 C)
3. Arduino compatible 
4. May be an encoder, rather than buttons

**Reply by VK6TT** [reply]
Here are some suggestions based on my extensive experience with JLPCB SMD assembly jobs (4-5 jobs per month, at least one of those is assembly):

1. Never place an order until one month has lapsed from the end of Chinese New Year (the returning workers presumably multi-skill a lot more in the first month and I found too many issues in that period.)
2. Never use 0402 or smaller parts - too many tombstones and mos-oriented parts and if you have to do any remedial work it's too hard
3. Really watch the diode orientation - this has bitten me so many times. On their part placement diagram ignore any markings on the diode bodies. + means anode and always anode. Until they get it wrong and then you're back to doing remedial work.
4. Check your footprints. Libraries do have errors. Go back and manually measure the various dimensions for each new footprint you use. Yes, I just had to scrap 50 boards because the SOIC16 pad layouts were wrong. Somehow I had two SOIC16 footprints in my libraries and one was good, one was bad. Murphy!
5. Use through hole parts when exact placement matters. First run I did with SMD switches like the W1209 boards had enough variation in placement that I had to re-work every case I had to suit. Next time I used through hole switches and no issues. 
6. BE VERY CAREFUL when selection parts. There are PNP transistors confusingly labelled as NPN/PNP. Check the datasheet to be sure you're getting a PNP or NPN. Just replaced 60 of those pesky PNP's yesterday. Again!
7. Don't expect any meaningful recompense if there is a mistake. Your warranty ends the moment you make payment. I gave up even asking.

When it all happens correctly it's very pleasing and cost effective.  But a painful learning curve.

**Reply by TG9541** [reply]
Hmmm... I'm not so sure about the Arduino thing (see above :-) ) but dimensions, load (type, wattage), display, connectors , connectivity... all would be interesting to know.

**Reply by TG9541** [reply]
@VK6TT thanks Richard, that's good advice! I already leaned about the orientation thing.  Are you using KiCad, too?

0402 - well, I think taht 'll have to give it a try - too much layout work went into that. Did you check your SMD pad dimensions?

**Reply by scientistnobee** [reply]
> 
> 
> Hmmm... I'm not so sure about the Arduino thing (see above :-) ) but dimensions, load (type, wattage), display, connectors , connectivity... all would be interesting to know.

Connection: MicroUSB/USB type-C input, wattage : 5W is enough for my application, Display: happy with current LEDs, connectivity: Not required

**Reply by scientistnobee** [reply]
> 
> 
> Here are some suggestions based on my extensive experience with JLPCB SMD assembly jobs (4-5 jobs per month, at least one of those is assembly):
> 
>     1. Never place an order until one month has lapsed from the end of Chinese New Year (the returning workers presumably multi-skill a lot more in the first month and I found too many issues in that period.)
> 
>     2. Never use 0402 or smaller parts - too many tombstones and mos-oriented parts and if you have to do any remedial work it's too hard
> 
>     3. Really watch the diode orientation - this has bitten me so many times. On their part placement diagram ignore any markings on the diode bodies. + means anode and always anode. Until they get it wrong and then you're back to doing remedial work.
> 
>     4. Check your footprints. Libraries do have errors. Go back and manually measure the various dimensions for each new footprint you use. Yes, I just had to scrap 50 boards because the SOIC16 pad layouts were wrong. Somehow I had two SOIC16 footprints in my libraries and one was good, one was bad. Murphy!
> 
>     5. Use through hole parts when exact placement matters. First run I did with SMD switches like the W1209 boards had enough variation in placement that I had to re-work every case I had to suit. Next time I used through hole switches and no issues.
> 
>     6. BE VERY CAREFUL when selection parts. There are PNP transistors confusingly labelled as NPN/PNP. Check the datasheet to be sure you're getting a PNP or NPN. Just replaced 60 of those pesky PNP's yesterday. Again!
> 
>     7. Don't expect any meaningful recompense if there is a mistake. Your warranty ends the moment you make payment. I gave up even asking.
> 
> 
> When it all happens correctly it's very pleasing and cost effective. But a painful learning curve.

Thanks, Richard, All these are valid points and great learnings.

**Reply by TG9541** [reply]
> Connection: MicroUSB/USB type-C input, wattage : 5W is enough for my application, Display: happy with current LEDs, connectivity: Not required

I don't think that I can do that - the USB-C specs look frightening...

**Reply by krikoman** [reply]
> 
> 
> > Most relays are good for at least 5A but either way the relay would simple burn out rather than drop in and out.
> > Are you powering your heater from the same power supply as the W1209? If this is the case, then probably what's happening is the relay switches, the load is switched and the voltage drops, the relay drop out, and drops the load, and it then cycles.
> > What size power supply are you using, do you have drawing of the circuit.
> > Also, depending on you heater element, there can be a massive inrush current, until it gets warmed up, a large load will pull the voltage down.
> 
> Yes, I am powering W1209 with the same source. My heater is having 3.6 Ohms. My power supply is a 5V battery bank. So the current draw is more than one ampere. The power bank's voltage is dropping to 3.6V when relay is on. So it's struggling. Is there anyway to solve this issue? Like shorting the diode or the resistor before the voltage regulator.

The easiest way is to use a "bigger" ( higher amp output) but that will depend on if you have a supply close to it.
What are you warming up? and how hot do you need it?
I tried a number of heating elements, but ended up using a USB Cup heater, who knew they existed?!?, anyhow it warms to 46 C and only pulls 0.6A so I can power it and the controller from 5V 1A supply.

**Reply by krikoman** [reply]
> 
> 
> Interesting discussions are going on here - there seems to be a real need for hackable thermostats ;-)
> @krikoman I've never heard of the CMS8S588 but it sounds like yet another STM8S003F3-look-alike. Can you post a photo with a close-up of the MCU? Thanks!
> 

Hope this works
![Capture](https://user-images.githubusercontent.com/12865684/107931905-d4157e80-6f74-11eb-82ee-ca8a506a6274.PNG)

**Reply by krikoman** [reply]
here's the board.
![P1010549](https://user-images.githubusercontent.com/12865684/107932533-9533f880-6f75-11eb-8319-dc9ce0952199.JPG)

**Reply by scientistnobee** [reply]
@krikoman I need up to 70 degree of heat. I am using extruder cartridge from a 3D-printer. Thanks for sharing the photos. Did you find a suitable 5V relay to test to see if it works at 5V? I used the below relay and it is much bigger than required. I couldn't find a suitable relay in Mouser, Farnell, digikey and RS. I don't know where and how to look for these. 
https://ie.rs-online.com/web/p/non-latching-relays/6839188/

**Reply by krikoman** [reply]
When you say bigger, do you mean physical size? Or it's a 10A relay?
If you are using cartridge heaters, they tend to have a high inrush current which will pull the voltage down, it can take 10's of seconds, with a reasonable power supply to get over this inrush, until the heater warms up, until the current drops to reasonable level.
Is the heater a 5V unit?

**Reply by scientistnobee** [reply]
Bigger in physical size. No the heater is not supposed to work with 5V. It is originally intended to work at 12V.

**Reply by scientistnobee** [reply]
> 
> 
> > Connection: MicroUSB/USB type-C input, wattage : 5W is enough for my application, Display: happy with current LEDs, connectivity: Not required
> 
> I don't think that I can do that - the USB-C specs look frightening...

I understand. Micro USB should be enough. I sent a DM on hackaday.io, could you please see that

**Reply by krikoman** [reply]
> 
> 
> Bigger in physical size. No the heater is not supposed to work with 5V. It is originally intended to work at 12V.

https://www.reichelt.com/at/en/dil-miniature-relay-hjr-4102-5-v-1-changeover-contact-5-a-hjr-4102-l-5v-p101950.html?&trstct=pol_0&nbc=1

HJR-4102E Form A

Mine came with the sensor, oblivious. dimensions are 15L x 12W x 13H
I see your difficulty finding them though.

FYI my 12VDC 30W cartridge heater pulled 1.275A @ 5V The 5V controller needs 9.3mA (Relay Off) and 68.2mA relay on.
Your 12V relay will require more current.
So with a 12V 30W heater you'll need ~1.3A supply You could put two in series, to drop the current, but not sure what that might do to the heat output.

**Reply by krikoman** [reply]
>  I couldn't find a suitable relay in Mouser, Farnell, digikey and RS. I don't know where and how to look for these.

Try this https://www.ebay.co.uk/itm/5V-Subminiature-PCB-Signal-Relay-1A-SPDT-HFD41/283642560171?hash=item420a69feab:g:EWsAAOSwmFVdo5iR

**Reply by scientistnobee** [reply]
Thanks for the links @krikoman very useful.

**Reply by scientistnobee** [reply]
Hello @krikoman I was trying to order the Relay you suggested but it seems that the pinout is not compatible with the original relay. Are you sure about this relay that it will work as drop in replacement to original 12V relay.

**Reply by krikoman** [reply]
Unfortunately I don't know for certain, as my board come with a 5V relay, so I've never needed to replace it.

**Reply by scientistnobee** [reply]
I ordered the below relay from ebay. Let's see how it works. Its impossible to find these on regular component suppliers such as RS, Farnell, Mouser, Digi-key. Or I might be too bad at searching for these. 

https://www.ebay.ie/itm/5V-Mini-Power-Songle-Relay-SPDT-5-Pins/261993762040?ssPageName=STRK%3AMEBIDX%3AIT&_trksid=p2057872.m2749.l2649

**Reply by krikoman** [reply]
> Interesting discussions are going on here - there seems to be a real need for hackable thermostats ;-)
> @krikoman I've never heard of the CMS8S588 but it sounds like yet another STM8S003F3-look-alike. Can you post a photo with a close-up of the MCU? Thanks!

Hi, I've got a bit  more info on my 5V version chip, though it wasn't easy to find, even after I'd found it the first time!!
It's in Chinese too, just for more fun.
Obviously, the chip might actually be a knock-off of the product below.


_CMS8S5880
CMS8S5880 is an enhanced Flash MCU with 1T 8051 core, working voltage 2.1V-4.5V, main frequency up to 24MHz, I/O up to 18, ADC, PWM, 2 UART, 1 SPI, 1 I2C, sleep state
The lower quiescent current is less than 6uA, and it can work from -40 to 105℃. Packages such as TSSOP20/QFN20 are available.
product: General purpose microcontroller, 8-bit microcontroller, CMS8S series_

Product sheet below.

[CMS8S5880-TSSOP20.pdf](https://github.com/TG9541/stm8ef/files/6035498/CMS8S5880-TSSOP20.pdf)

**Reply by scientistnobee** [reply]
Hi everyone, 

I got the same 5V w1209 module suggested by @krikoman today in mail. This is again not working with my heating cartridge that is pulling 1.25 Amps. Its not working with both battery pack and the wall-plug power source (Raspberry pi). So, I took the relay from this 5V module and soldered to 12V module and that 12V module worked without any issues. I noticed the same chip on 5V module as posted by @krikoman and it has a 3.3V regulator. I think this is the reason its not working with my catridge, but can't say for sure. 

Hi @krikoman where did you get the coffee heating pad. Could you please give the link?

**Reply by krikoman** [reply]
Heater pad was here but mines already dead. I think it carbon fibres across a couple of bits f copper, see below. It stopped getting hot!
https://www.amazon.co.uk/dp/B08PVGHJC1/ref=pe_2515581_431334481_fb_remail%5Fprodpg
The 3.3.V regulator doesn't care about the cartridge, the cartridge supply is through the contacts of the relay and the power supply, so the 3.3V shouldn't really come into it. But at least its working now.

This is the cup heater
![IMG_20210223_181713](https://user-images.githubusercontent.com/12865684/110134439-4edbf780-7dc5-11eb-8d0d-afa6ecd07ec1.jpg)

This is what's inside.
![IMG_20210223_181745](https://user-images.githubusercontent.com/12865684/110134463-526f7e80-7dc5-11eb-8d3a-b76136c53839.jpg)

What you can do with a 5V supply is to wire two 12v 30w heaters in series, which will keep the current down. This was what I've done when the cup heater gave up!

**Reply by krikoman** [reply]
Do you have a multimeter you can use?
It might be interesting to understand what's happening with some actual readings.

**Reply by scientistnobee** [reply]
I have a Multimeter, a good one with me.

**Reply by krikoman** [reply]
It might be nice to record the voltage values across  the supply cables when the relay is off and again when on for both the power supplies you have.
This would have been good for the 5V unit before you removed the relay too, but that's too late now, unless you bought two.

Current readings might be good too.

**Reply by scientistnobee** [reply]
I have only one 5V module from which I already taken out the relay. So, won't be useful. But I i have two 12V modules each with 5V relays (one from 5V W1209, another one big OMRAN relay). I tested the voltage when the relay is on and off with both power supplies. My finding are following. 


1. Wall-plug: voltage across power lines: 5.15 V both when relay is on and off), voltage at shorted voltage regulator is 3.6V and 4.3V when the relay is on and off respectively. 
2.  Power bank: voltage across power lines: 4.4 V when relay is on and 5.06 when relay is off), voltage at shorted voltage regulator is 3.0V and 4.1V when the relay is on and off respectively.

**Reply by krikoman** [reply]
Is this with the load connected too, or just the unit itself?
Thanks for the info.

**Reply by scientistnobee** [reply]
Yes, these values are with load connected to the relay. I edited the above comment when the power bank is connected. There was a typo in the values when the relay is on for the case of power bank.

**Reply by krikoman** [reply]
Nice one, cheers, for the info.

**Reply by scientistnobee** [reply]
I set-up the W1209 for 30 degrees, and hysteresis of 0.1. I measured the temperature with another thermistor and I got the below plot. So the maximum is around 40 and minimum 33 degrees. Anyone experienced this. I think reason for this kind of variance is due to the higher hysteresis. Any suggestions to improve the performance, i.e. make the peaks and valleys within 1 or 2 degrees.

![image](https://user-images.githubusercontent.com/7974878/110535790-9a213d80-8118-11eb-90e9-d7f18be22652.png)

My set-up

![image](https://user-images.githubusercontent.com/7974878/110547131-6994d000-8127-11eb-8d4f-46308313082a.png)

**Reply by krikoman** [reply]
Where's you relay sensor in relation to the heater, what sort of space are you trying to warm up.

My sensor is on the same heat sink as the heater, this is all enclosed in a smallish volume of 2,000 cm^3, and it's insulated from the outside.
My hysteresis on the devices is 0.5C and the temp. varies by around 0.75C depending on the outside temp. This could be improved by greater insulation, I think.

**Reply by krikoman** [reply]
OK just zoomed into the photo, so I can see what you've got set up.
If you insulated to block it might help smooth things out. It does seem a bit strange though the large range on something so small.

Do you have two probes on the controller?

Do you know what the temperature rating the cartridge is? It might be this is getting really hot very quickly and by the time it's sensed and witched off it's already way over the setpoint.

I tried to get one around the target temperature, but started off with one that was about 3 times over and it got hot too quickly. This meant the relay was switching more often than was required.

**Reply by scientistnobee** [reply]
Thanks for the suggestions. Second sensor is just laying around. It was not connected to the board. The cartridge sensor in the current set-up is 12V, 40W and resistance 4 Ohms. Now I changed the cartridge to 24V, 40W, resistance 12 Ohms. I got the following graph for 32 degrees set temperature, it is better than previous set-up with less resistance cartridge. Moreover, 5V w1209 module worked without any issue with this 24V, 40W cartridge. As you suggested, I will increase the resistance further to 20ohms with 24V, 30W cartridge. I hope I will get more flat response. 

![image](https://user-images.githubusercontent.com/7974878/110808260-b8537e80-827b-11eb-80d1-337dbee27e3e.png)

**Reply by scientistnobee** [reply]
The above graph was with a new module where I failed to check the hysteresis setting, which was kept at 2 degrees. Now, I am getting better results with hysteresis of 0.1 and P4 (correction term) of 3 degrees for a temperature setting of 37 degrees. All these with a cartridge of 24v, 40W. You can see +-1 degree change from 37 degrees (only see lat 3-4 cycles). I think I can improve this further by using a 24V, 30W cartridge, which is on the way. 

![new_catridge_37_correct_degrees](https://user-images.githubusercontent.com/7974878/110934177-96163b00-8325-11eb-8ab9-9ccd7029f5b0.png)

**Reply by scientistnobee** [reply]
The correction term has to change at different temperatures, for example for 55 degrees, P4 is  around 6.7 degrees.

![new_catridge_55_correct_degrees](https://user-images.githubusercontent.com/7974878/110934453-f0af9700-8325-11eb-9519-c6c1e72bd85c.png)

**Comment by scientistnobee** [comment]
Thanks for your prompt reply. Is it okay to short input and output of the 12V regulator? Is there any link where I can buy genuine boards?

**Reply by TG9541** [reply]
Genuine boards are rare these days, and I was't able to find a reliable source (I tried!). I have a small stock, though, and maybe I can send one your way. Where do you live?

**Comment by TG9541** [comment]
I think that shorting in and out should be OK. The thing I proposed would result in a slightly higher supply voltage.
Genuine boards have become as rare as hen's teeth. I asked several vendors but the supply side appears to be dead. The last time I got a board with an STM8S003 controller was nearly two years ago. Today it's almost easier to design your own board with KiCad and have JLCPCB assemble it for you.

**Comment by scientistnobee** [comment]
I did some experiments with the module by just decreasing the supply voltage. I think circuit is working as it is with 5v without any shorting of voltage regulator, except the relay. What I am thinking is can I turn on 12v relay with 5v supply voltage, just by changing 2k resistor at the base of the transistor? Any suggestions?

**Reply by TG9541** [reply]
I would try it by changing just the relay. Measure the current and check if it's within the specs.

**Reply by scientistnobee** [reply]
Okay. I have potentiometer now, but no 5v relay yet. Also, I have to search compatible relay.

**Reply by TG9541** [reply]
It's best to be careful about the relay specs. A relay may switch-on at a certain voltage when it's cold but won't do that when it's warm (relays get warm while active).

**Comment by TG9541** [comment]
@scientistnobee I'll first convert this issue into a [discussion thread](https://github.com/TG9541/stm8ef/discussions)

**Comment by scientistnobee** [comment]
Hi, if you don't mind me asking a silly question, I would like to know what is the role of R3 between the two diodes?

**Reply by TG9541** [reply]
To be honest I have no idea why the designer of the board decided to apply a 10R resistor in the 12V supply since the value of C5 is much too small for R3 to have a meaningful effect. The supply current is easily 0.15A, and that means that it leads to an additional power consumption of 0.2W.

**Comment by scientistnobee** [comment]
Thanks for your reply, I have several other issues with understanding the design choices. I wonder about this resistor the most. Also, while you are here may I ask, if there exist anyway to code this with Arduino IDE?

**Reply by TG9541** [reply]
It's certainly possible to find a solution with the Arduino IDE but I don't know how complicated it is to set such a thing up and how good the support for STM8 actually is. This is maybe a starting point: 

https://github.com/stm32duino/Arduino_Core_STM8

You may also have a look at something that might be close to what you need but that will bring you in touch with a different technology:

https://github.com/TG9541/W1209

**Reply by scientistnobee** [reply]
Thanks for the links. I was aware of eFORTH, but I afraid of learning a new syntax

**Reply by TG9541** [reply]
The syntax is easy: there is none. The only thing is that it's "1 2 +" instead of "1 + 2". Function calls are "5 setADC" instead of "setADC(5)". This looks a bit strange but it takes just about 5 minutes to get familiar with.

There are also cases where the "C-way" looks awkward in comparison: instead of writing "f3(f2(f1(5)))" for chained functions in Forth it's just a chain of functions "5 f1 f2 f3" :-)

**Reply by scientistnobee** [reply]
Thanks for the nudge, I will give it a try and get back to you.

**Comment by VK6TT** [comment]
Hi Mark,

sounds like a variant of the original W1209 boards which started Thomas 
on his journey. That means you will have to change the IC out for a 
genuine STM8 device.  There is a page in the wiki detailing how to avoid 
a clone.

Scary for the first timer but not that hard in practice. I couldn't 
access your order details so I couldn't visually look at the board for you.

Regards

Richard

**Reply by krikoman** [reply]
I thought it might be. It works great for what I need it for, though it would be better with more data, temps and relay operations.
I bought a 12V one (because it came with a case and I hadn't ordered one for my 5V version) but that has no markings on the IC at all, it was interesting to try it with my PSU, it worked down to 4.4V, but the relay wouldn't pull in (as expected) it needed 7+ volts to pull in the relay.
I put some photos up the thread, if you want more let me know.
I'm amazed how well these things work for how little they cost.
